internal/reconcilers/cluster.go

```diff
@@ -0,0 +1,154 @@
+type hardwareIPGetter interface {
+	NextAvailableHardwareID(ctx context.Context) (string, error)
+	GetHardwareIP(ctx context.Context, uuid string) (string, error)
+}
+
+// clusterReconcileContext implements ReconcileContext by reconciling TinkerbellCluster object.
+type clusterReconcileContext struct {
+	ctx               context.Context
+	tinkerbellCluster *infrastructurev1alpha3.TinkerbellCluster
+	patchHelper       *patch.Helper
+	hardwareIPGetter  hardwareIPGetter
+	cluster           *clusterv1alpha3.Cluster
+	log               logr.Logger
+	client            client.Client
+}
+
+// New builds a context for cluster reconciliation process, collecting all required
+// information.
+//
+// If unexpected case occurs, error is returned.
+//
+// If some data is not yet available, nil is returned.
+func (crcc *ClusterReconcileContextConfig) New(namespacedName types.NamespacedName) (ReconcileContext, error) {
+	crc := &clusterReconcileContext{
+		log:               crcc.Log.WithValues("tinkerbellcluster", namespacedName),
+		ctx:               context.Background(),
+		tinkerbellCluster: &infrastructurev1alpha3.TinkerbellCluster{},
+		hardwareIPGetter:  crcc.HardwareIPGetter,
+		client:            crcc.Client,
+	}
+
+	if err := crc.client.Get(crc.ctx, namespacedName, crc.tinkerbellCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+
+		return nil, fmt.Errorf("getting TinkerbellCluster: %w", err)
+	}
+
+	crc.log = crc.log.WithName(crc.tinkerbellCluster.APIVersion)
+
+	patchHelper, err := patch.NewHelper(crc.tinkerbellCluster, crc.client)
+	if err != nil {
+		return nil, fmt.Errorf("initializing patch helper: %w", err)
+	}
+
+	crc.patchHelper = patchHelper
+
+	cluster, err := util.GetOwnerCluster(crc.ctx, crc.client, crc.tinkerbellCluster.ObjectMeta)
+	if err != nil {
+		return nil, fmt.Errorf("getting owner cluster: %w", err)
+	}
+
+	if cluster == nil {
+		crc.log.Info("OwnerCluster is not set yet. Requeuing...")
+
+		return nil, nil
+	}
+
+	crc.cluster = cluster
+
+	if util.IsPaused(crc.cluster, crc.tinkerbellCluster) {
+		crc.log.Info("TinkerbellCluster or linked Cluster is marked as paused. Won't reconcile")
+
+		return nil, nil
+	}
+
+	return crc, nil
+}
+
+// Reconcile implements ReconcileContext interface by ensuring that all TinkerbellCluster object
+// fields are properly populated.
+func (crc *clusterReconcileContext) Reconcile() error {
+	if crc.tinkerbellCluster.Spec.ControlPlaneEndpoint.Host == "" {
+		id, err := crc.hardwareIPGetter.NextAvailableHardwareID(crc.ctx)
+		if err != nil {
+			return fmt.Errorf("getting next available hardware: %w", err)
+		}
+
+		ip, err := crc.hardwareIPGetter.GetHardwareIP(crc.ctx, id)
+		if err != nil {
+			return fmt.Errorf("getting hardware IP: %w", err)
+		}
+
+		crc.log.Info("Assigning IP to cluster", "ip", ip, "clusterName", crc.tinkerbellCluster.Name)
+
+		crc.tinkerbellCluster.Spec.ControlPlaneEndpoint.Host = ip
+	}
+
+	// TODO: How can we support changing that?
```

Comment:
It's a bit more complicated in our case currently because we don't have a load balancer (yet), I don't think we have a good way (other than making assumptions related to the bootstrap provider).